src/runtime_src/core/common/memaccess.cpp

```diff
@@ -0,0 +1,199 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#define XRT_CORE_COMMON_SOURCE
+#include "memaccess.h"
+
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+#include "memalign.h"
+#include "query_requests.h"
+#include "utils.h"
+#include "unistd.h"
+
+// TODO add tag in here for alter use...
+struct mem_bank_t {
+  uint64_t m_base_address;
+  uint64_t m_size;
+  int m_index;
+  uint8_t m_type;
+  mem_bank_t (uint64_t addr, uint64_t size, int index, uint8_t type) : m_base_address(addr), m_size(size), m_index(index), m_type(type) {}
+};
+
+static std::vector<mem_bank_t>
+get_ddr_banks(const xrt_core::device* device)
+{
+  std::vector<mem_bank_t> banks;
+
+  auto mt_raw = xrt_core::device_query<xrt_core::query::mem_topology_raw>(device);
+  auto map = reinterpret_cast<const mem_topology*>(mt_raw.data());
+
+  for( int i = 0; i < map->m_count; i++ ) {
+      // If a memory bank is in use and is not a streaming bank emplace it into the bank list
+      if( map->m_mem_data[i].m_used && map->m_mem_data[i].m_type != MEM_STREAMING )
+          banks.emplace_back( map->m_mem_data[i].m_base_address, map->m_mem_data[i].m_size*1024, i, map->m_mem_data[i].m_type );
+  }
+
+  std::sort (banks.begin(), banks.end(),
+              [] (const mem_bank_t& a, const mem_bank_t& b) {return (a.m_base_address < b.m_base_address);});
+
+  return banks;
+}
+
+/*
+  * validate_address_and_size()
+  *
+  * Sanity check the user's Start Address and Size against the mem topology
+  * If the start address is 0 (ie. unspecified by user) change it to the first available address
+  * If the size is 0 (ie. unspecified by user) change it to the maximum available size
+  * Fill the vector with the available banks
+  * Set the iterator to the bank containing the start address
+  */
+static void
+validate_address_and_size(xrt_core::device* device, uint64_t& start_addr, uint64_t& size,
+            std::vector<mem_bank_t>& vec_banks, std::vector<mem_bank_t>::iterator& start_bank) 
+{
+  // Validate the memory banks
+  vec_banks = get_ddr_banks(device);
+  //Find the first memory bank with valid size since vec_banks is sorted
+  auto valid_bank =  std::find_if(vec_banks.begin(), vec_banks.end(),
+              [](const mem_bank_t item) {return (item.m_size);});
+
+  if (valid_bank == vec_banks.end()) {
+    std::cerr << "ERROR: Couldn't find valid memory banks\n";
+    throw xrt_core::error(std::errc::operation_canceled);
+  }
```

Comment:
```suggestion
  if (valid_bank == vec_banks.end())
    throw xrt_core::error(std::errc::operation_canceled, "Couldn't find valid memory banks");
```
You are in coreutils now,  you cannot let any code write to std::cout or std::err.   Your code is now accessible to library code and it must not write to stdout or err.

Please fix all instances.